crypto/x509/utils/src/lib.rs

```
@@ -0,0 +1,292 @@
+    /// Create an iterator over a string which contains Pem objects
+    fn into_pem_iter(self) -> PemStringIter<Self>;
+}
+
+/// Anything which can be referenced as a byte slice gets into_iter_pem()
+impl<T: AsRef<[u8]>> PemStringIterable for T {
+    fn into_pem_iter(self) -> PemStringIter<T> {
+        PemStringIter::new(self)
+    }
+}
+
+/// An iterator of [`X509Certificate`] objects over a slice of [`Pem`] objects.
+pub struct X509CertificateIter<'a> {
+    pem_slice: &'a [Pem],
+    offset: usize,
+}
+
+impl<'a> X509CertificateIter<'a> {
+    fn new<T: AsRef<[Pem]>>(pems: &'a T) -> Self {
+        Self {
+            pem_slice: pems.as_ref(),
+            offset: 0,
+        }
+    }
+}
+
+impl<'a> Iterator for X509CertificateIter<'a> {
+    type Item = X509Certificate<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.offset == self.pem_slice.len() {
+            None
+        } else {
+            X509Certificate::from_der(&self.pem_slice[self.offset].contents)
+                .map(|(_, x509)| {
+                    self.offset += 1;
+                    x509
+                })
+                .ok()
+        }
+    }
+}
+
+/// A trait used to monkey-patch an X509Certificate parsing iterator over a
+/// slice of [`Pem`] objects.
+pub trait X509CertificateIterable {
+    fn iter_x509(&self) -> X509CertificateIter;
+}
+
+/// Anything which can be referenced as a Pem slice gets a method to iterate
+/// certificates, and verify it is a well-formed single-path certificate chain.
+impl<T: AsRef<[Pem]>> X509CertificateIterable for T {
+    fn iter_x509(&self) -> X509CertificateIter {
+        X509CertificateIter::new(self)
+    }
+}
+
+/// A trait used to monkey-patch an X509Certificate chain verifier onto a slice
+/// of X509Certificate objects.
+pub trait X509CertificateChain {
+    /// Verify the chain (checks validity, signatures, and CA extension of each
+    /// element)
+    fn verify_chain(&self) -> bool;
```

it might be better to return an error with Display impl here?